File: packages/capacitor-actito-loyalty/android/src/main/java/com/actito/loyalty/capacitor/ActitoLoyaltyPlugin.kt
package com.actito.loyalty.capacitor

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.actito.Actito
import com.actito.ActitoCallback
import com.actito.loyalty.ktx.loyalty
import com.actito.loyalty.models.ActitoPass

@CapacitorPlugin(name = "ActitoLoyaltyPlugin")
public class ActitoLoyaltyPlugin : Plugin() {
    @PluginMethod
    public fun fetchPassBySerial(call: PluginCall) {
        val serial = call.getString("serial") ?: run {
            call.reject("Missing 'serial' parameter.")
            return
        }

        Actito.loyalty().fetchPassBySerial(serial, object : ActitoCallback<ActitoPass> {
            override fun onSuccess(result: ActitoPass) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun fetchPassByBarcode(call: PluginCall) {
        val barcode = call.getString("barcode") ?: run {
            call.reject("Missing 'barcode' parameter.")
            return
        }

        Actito.loyalty().fetchPassByBarcode(barcode, object : ActitoCallback<ActitoPass> {
            override fun onSuccess(result: ActitoPass) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun present(call: PluginCall) {
        val json = call.getObject("pass") ?: run {
            call.reject("Missing 'pass' parameter.")
            return
        }

        val pass: ActitoPass = try {
            ActitoPass.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        val activity = activity ?: run {
            call.reject("Cannot present a pass without an activity.")
            return
        }

        Actito.loyalty().present(activity, pass)
        call.resolve()
    }
}

File: packages/capacitor-actito-loyalty/ios/Plugin/ActitoLoyaltyPlugin.swift
import ActitoKit
import ActitoLoyaltyKit
import Capacitor
import Foundation

@objc(ActitoLoyaltyPlugin)
public class ActitoLoyaltyPlugin: CAPPlugin {
    private var rootViewController: UIViewController? {
        get {
            UIApplication.shared.delegate?.window??.rootViewController
        }
    }

    @objc func fetchPassBySerial(_ call: CAPPluginCall) {
        guard let serial = call.getString("serial") else {
            call.reject("Missing 'serial' parameter.")
            return
        }

        Actito.shared.loyalty().fetchPass(serial: serial) { result in
            switch result {
            case let .success(pass):
                do {
                    call.resolve([
                        "result": try pass.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func fetchPassByBarcode(_ call: CAPPluginCall) {
        guard let barcode = call.getString("barcode") else {
            call.reject("Missing 'barcode' parameter.")
            return
        }

        Actito.shared.loyalty().fetchPass(barcode: barcode) { result in
            switch result {
            case let .success(pass):
                do {
                    call.resolve([
                        "result": try pass.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func present(_ call: CAPPluginCall) {
        guard let json = call.getObject("pass") else {
            call.reject("Missing 'pass' parameter.")
            return
        }

        let pass: ActitoPass

        do {
            pass = try ActitoPass.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                call.reject("Cannot present a pass with a nil root view controller.")
                return
            }

            Actito.shared.loyalty().present(pass: pass, in: rootViewController)
        }
    }
}
